### PR TITLE
Add support for Python 3.11

### PR DESCRIPTION
**Pull request by parasj**
Changes:
* Update Dockerfile to Python 3.11
* Test pytest on CI with Python 3.11
* Move development dependencies to separate file to improve compatibility

### CODE REVIEW

File: CONTRIBUTING.rst [hunk 1]
```diff
@@ -41,7 +41,7 @@ Before you submit a pull request, make sure to complete the following steps:
 
    .. code-block:: console
 
-      $ pip install black pytype autoflake
+      $ pip install -r requirements-dev.txt
```
Comment (phi-line) [inline CONTRIBUTING.rst:44]: what about:

```shell
poetry install --only dev
```

File: pyproject.toml [hunk 2]
```diff
@@ -12,7 +12,7 @@ readme = "README.md"
 include = ["profiles/*"]
 
 [tool.poetry.dependencies]
-python = ">=3.7.1,<3.11"
+python = ">=3.7.1,<4.0"
```
Comment (phi-line) [inline pyproject.toml:15]: is `<4.0` really the best way to capture this? There's no guarantee of a `4.0` release, and it seems like the next releases are `3.12` and `3.13`:
https://docs.python.org/3.12/whatsnew/3.12.html

**Review by phi-line — COMMENTED** [review]
I'm a little confused with the use of `poetry` in this project. I would opt for maintaining the `pyproject.toml` with groups for `dev` and `gateway` instead of using a combination of `poetry`, `pip`, and `requirements.txt`s.

File: CONTRIBUTING.rst (hunk 1 above)
Comment (parasj) [reply CONTRIBUTING.rst:44]: We have an issue that Poetry requires Python version compliance with all dev dependencies and will not allow specification of a different Python version.

There's an in progress PR to add support for this, but it won't be production-ready for a while: https://github.com/python-poetry/poetry/pull/5740

Several dev dependencies have not been updated for Python 3.11, most notably pytype which requires python `>=3.7.1<3.11`.

Until that PR lands, an alternative solution is to extract development CLI dependencies into another file. Thoughts?

File: pyproject.toml (hunk 2 above)
Comment (parasj) [reply pyproject.toml]: Agreed, updated to 3.12

**Review by phi-line — APPROVED** [review]

**Comment by phi-line** [comment]
> Package management is a bit confusing

Thanks for the explanation!

**Comment by parasj** [comment]
@phi-line Thanks for the details feedback on these PRs. I just added you as a contributor to the project. Are you on the Skyplane Slack? https://join.slack.com/t/skyplaneworkspace/shared_invite/zt-1cxmedcuc-GwIXLGyHTyOYELq7KoOl6Q

**Pull request merged**